File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 6 - Polynomial Regression/PolynomialRegression.py
#Polynomial Regression
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing Dataset
dataset=pd.read_csv('Position_Salaries.csv')
X=dataset.iloc[:,1:2].values
Y=dataset.iloc[:,2].values

# Encoding categorical data
# Encoding the Independent variable
"""from sklearn.preprocessing import LabelEncoder,OneHotEncoder
labelencoder_X=LabelEncoder()
X[:,3]=labelencoder_X.fit_transform(X[:,3])
onehotencoder=OneHotEncoder(categorical_features=[3])
X=onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap
X=X[:,1:]

#Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,random_state=0)
"""
# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X =StandardScaler()
X_train=sc_X.fit_transform(X_train)
X_test=sc_X.transform(X_test)"""

# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg=LinearRegression()
lin_reg.fit(X,Y)

#Fittiong Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg=PolynomialFeatures(degree=4)
X_poly=poly_reg.fit_transform(X)
lin_reg_2=LinearRegression()
lin_reg_2.fit(X_poly,Y)

# Visualizing the linear Regression results
plt.scatter(X , Y , color = 'red')
plt.plot(X,lin_reg.predict(X),color = 'blue')
plt.title('Truth or Bluff(Linear Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Visualizing the linear Regression results
X_grid=np.arange(min(X),max(X),0.1)
X_grid=X_grid.reshape((len(X_grid) ,1))
plt.scatter(X , Y , color = 'red')
plt.plot(X,lin_reg_2.predict(poly_reg.fit_transform(X)),color = 'blue')
plt.title('Truth or Bluff(Linear Regression)')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Predicitng a new result with linear Regression
lin_reg.predict(6.5)

# Predicting a new result with polnomial Regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))

File: Machine Learning A-Z Template Folder/Part 2 - Regression/Section 6 - Polynomial Regression/PolynomialRegression.R
#Polynomial Regression

#importing the dataset

dataset=read.csv('Position_Salaries.csv')
dataset=dataset[,2:3]

# Encoding categorical data
dataset$State=factor(dataset$State,
                     levels=c('New York','California','Florida'),
                     labels=c(1,2,3))

# Spliting the dataset into the Training set and Test set
#install.packages('caTools')
#library(caTools)
#set.seed(123)
#training_set=subset(dataset,split==TRUE)
#test_set=subset(dataset,split==FALSE)

#  #Feature Scaling
#  training_set[,2:3]=scale(training_set[,2:3])
#  test_set[,2:3]=scale(test_set[,2:3])

#Fitting Linear Regression to the dataset

lin_reg=lm(formula = Salary ~ .,data = dataset)

#Fitting Polynomial Regression to the dataset

dataset$Level2=dataset$Level^2
dataset$Level3=dataset$Level^3
dataset$Level4=dataset$Level^4
poly_reg = lm(formula = Salary ~ .,data = dataset)

# Visualizing the Linear Regression results
library(ggplot2)
ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             colour='red')+
  geom_line(aes(x=dataset$Level,y=predict(lin_reg,newdata = dataset)),
            colour='blue')+
  ggtitle('Truth or Bluff (Linear Regression)')+
  xlab('Level') +
  ylab('Salary')
# Visualizing the Polynomial Regression results

library(ggplot2)
ggplot()+
  geom_point(aes(x=dataset$Level,y=dataset$Salary),
             colour='red')+
  geom_line(aes(x=dataset$Level,y=predict(poly_reg,newdata = dataset)),
            colour='blue')+
  ggtitle('Truth or Bluff (Polynomial Regression)')+
  xlab('Level') +
  ylab('Salary')

# Predicting a new resut with linear Regression
y_pred = predict(lin_reg,data.frame(Level = 6.5))

# Predicting a new Result with  Polynomial Regression
y_pred = predict(poly_reg,data.frame(Level = 6.5,Level2 = 6.5^2,Level3 = 6.5^3,Level4 = 6.5^4))
